Add FoldCase API

fbacher opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
Address Turkish-I problem (#19883)
<!--- Describe your change in detail here. -->
This is a large change:
Adds StringUtils::FoldCase to replace most uses of ToLower
Modifies ToLower & ToUpper to be more 'accurate' and give consistent results whether used as UTF-8 or wstring. This includes these three APIs using character tables based on the same current Unicode tables. Changed related APIs (EqualsNoCase, etc.) to use the new FoldCase, etc. Changed these APIS to use StringView and return String instead of pass by reference. 

Created modified CharacterSetEncoder. 
Changed all usages of ToLower/ToUpper, etc., to use the new APIs.
## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
The first priority was to have a proper fix for the Turkish-I problem
A secondary consideration was to improve Unicode support in general. For example, by performing case changes on Unicode code-points (4 bytes) instead of on single (utf8) or 16-bits (Windows) the quality of the those operations should improve.
<!--- If it fixes an open issue, please link to the issue here. -->
https://github.com/xbmc/xbmc/issues/19883
## How has this been tested?
<!--- Please describe in detail how you tested your change. -->

- Wrote new test cases using test strings from multiple languages, including Turkish and the offending Turkish "I".
- I did NOT write tests to measure improvement of case-changing by operating on Unicode chars versus only the first byte of a multi-byte utf8 sequence
- Passed all test case
- Wrote performance tests to check for no unexpected performance degradation 
- Ran Kodi under different languages, intentionally choosing most with multi-byte characters, some Western-European, Cyrillic based, Chinese, Japanese and Thai.
- Downloaded movie data from TMDb to create several thousand fake movie entries.
- Exercised Kodi manually as well as with a RandomTrailers addon I wrote
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Development and test platform is a 6/12 cpu Intel Ubuntu 22.10 system. 
Most tests already mentioned above. 
The only exposure I have to other platforms so far is through the PR builds where Windows and OSx have required some changes. 
## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
Ensures Kodi is usable in Turkish family of languages (2 or 3).
Hopefully improves character handling in multi-byte languages (since ToLower/ToUpper
and compare operations should be more accurate)
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), --> 
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
There should be little user-noticeable changes, only improvements in quality of character handling
## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)
Changing to use string_view and returning strings over using by reference will be, I think, a welcome change. Removing the old APIs will likely not be too popular. (ToLower/ToUpper and perhaps several more StringUtils calls). I can likely put the other APIs back but it will be even more bloated than before.
## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x ] I have updated the documentation accordingly
- [x ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x ] I have added tests to cover my change
- [x ] All new and existing tests passed

fbacher committed:
Use FoldCase for xbmc/utils

Switch all of the apprpriate calls to ToLower in xbmc/utils/... to use
    FoldCase instead

fbacher committed:
Use FoldCase for root, addons, application

Use FoldCase for: xbmc, xbmc/addons and xbmc/application; as appropriate
in place of ToLower

fbacher committed:
Use FoldCase in xbmc/cores

Use FoldCase instead of ToLower in the appropriate situations in the
xbmc/cores subdirectory

fbacher committed:
Use Foldcase for filesystem, games, gulib, input

Replace the use of ToLower with FoldCase in the appropriate places for
xbmc/filesystem, xbmc/games, xbmc/guilib, xbmc/input

fbacher committed:
Use FoldCase for settings, video, windowing, windows

Use FoldCase instead of ToLower for the appropriate instances in the
xbmc/settings, xbmc/video, xbmc/windowing and xbmc/windows directories.

fbacher committed:
Use FoldCase for interfaces, media, music, network, peripherals

Use FoldCase instead of ToLower in the appropriate situations for
xbmc/interfaces,
xbmc/media, xbmc/music, xbmc/network and xbmc/peripherals

fbacher committed:
Change *NoCase functions to use FoldCase

In addition, add some more functions where performance can be improved,
such as EqualsNoCase in addition to CompareNoCase.
Also, added Equals since a length-check can avoid the comparison.
Changed APIs to use string_view instead of the various permutations of
string and char*

fbacher committed:
Use Foldcase for pictures, platform, playlists, programs, pvr

Use Foldcase instead of ToLower in the appropriate cases for
xbmc/pictures, xbmc/platform, xbmc/playlists, xbmc/programs
xbmc/pvr

fbacher committed:
ToUpper/ToLower

Modify API to be consistent with FoldCase:
- Pass by value string_value
- return string

Modify uses accordingly, change to use FoldCase or use
StringUtils::getCLocale()
as appropriate

fbacher commented:
Sorry to start yet another PR. I tried to make several large cleanups on earlier commits and had lots of grief. My attempt to rebase the old branch using a new one did not bode well, so I just switched to the new branch.

I need to look closely at the released code and see what patch was made to work around the Turkish-I problem. 
I will add some performance measurements that I made. There shouldn't be much left to do before submitting this.

fbacher commented:
I'm having trouble applying  the lovely formatting patches that Jenkins supplied. I was adding mail headers to the .patch files and applying them with am, but I have lost my prototype file for that and can't find the instructions. Now, when I try to apply the first patch via 'git am' or 'git apply' I get merge conflicts. But it seems that I was not getting those when I added headers.

Should I expect occasional merge conflicts? Should I use 'apply' or 'am'?

fbacher committed:
clang-format

Only clang-format changes.

fbacher committed:
FoldCase API

Kodi NOT useable until next commit. All but FoldCase related tests pass.

fbacher committed:
Add private iconv (UnicodeConverter)

CharSetConverter has several performance and functional limitations that
are fairly easy to remedy.
Will propose merging the changes into CharSetConverter once this is
complete.

Kodi and tests functional

fuzzard commented:
> Problem here resolved
> 
> I'm having trouble applying the lovely formatting patches that Jenkins supplied. I was adding mail headers to the .patch files and applying them with am, but I have lost my prototype file for that and can't find the instructions. Now, when I try to apply the first patch via 'git am' or 'git apply' I get merge conflicts. But it seems that I was not getting those when I added headers.
> 
> Should I expect occasional merge conflicts? Should I use 'apply' or 'am'?
> 
> Figured out how to format to use 'git am' again. And I did learn more about how this works and it is not quite what I expected. I ended up:
> 
> 1. Saving all of my commits as patch files
> 2. Resetting HEAD to the commit that the first formatting patch here applies to
> 3. Apply the formatting patch
> 4. Force push & wait for build results and new formatting patches
> 5. Download the formatting patches, throwing away the old ones
> 6. Apply subsequent saved commit patch files until I get to the one corresponding to  next formatting patch here
> 7. Go to 3 (feels like FORTRAN)
>    Actually, I applied patches until I hit an error and then reset HEAD (as needed) to omit the borked changes and force pushed them for a build.
> 
> Hopefully I will remember this next time...

That's a long winded way, save yourself the hassle.

Do a git rebase and edit each commit that has a diff. During the rebase apply the appropriate diff, git add, git commit --amend, and then git rebase --continue. Repeat the patch/add/amend/continue for each commit

fbacher commented:
**Overview**

The primary aim of this PR is to properly fix problems introduced by using multi-lingual, multibyte handling in Kodi. The prime example is Kodi being unusable when the language is set to Turkish due to that language's peculiar rules on changing the case of the Turkish letter I, all four of them.

The solution is to replace "ToLower" with "FoldCase" in most occurrences in Kodi. FoldCase essentially does the same job as ToLower, except it ignores the locale, producing consistent results. This is a simple foldcase implementation, which means that it doesn't realize that a german sharfes-S (looks like a script B) should be converted to "ss". Fortunately Kodi doesn't appear to use such characters as keywords.

In all, about 450 uses of ToLower were changed to FoldCase. Almost always it is easy to spot when FoldCase should be used and when not. 

- If the string is used as a "word" to be displayed to a user, then us ToLower (unless you are informing the user what actual keyword value is used by the system).
- If the string is used as a key to a map, database index, string search, 
  etc. then FoldCase is used.
  
Implementing Foldcase involved using CaseFolding tables from Unicode.org,
which develops ICU4C. A fairly simple program was written to convert this table to a two-level lookup table in order to compress the large but sparse table. These tables use UTF-32 (aka UCS-4). Due to C++ STL shortcomings with processing Unicode, almost all FoldCase operations are done in UTF-32.

**UnicodeConverter**
Conversion between UTF-8, WCHAR and UTF-32 became a problem. C++ provides wstring_convert for converting to different encodings, but there is no converter to/from wstring to u32string. Further, several converters are deprecated but without a replacement. In addition, error handling is limited to replacing the entire string with a fixed message. I then tried to use CharSetConverter, based on iconv, but it also had problems:

- CharSetConverter was not initialized when StringUtils needed it, very early  in startup. CharSetConverter became available after about 50,000 calls to StringUtils for conversion. I spent a fair amount of time  trying to resolve without success. I have learned a bit since then so I could probably solve it now.
- CharSetConverter uses exclusive locks to prevent a converter being used by two threads at once. This may have contributed to the initialization issue. I did no measurements, but locks always slow things down. Further, the serialized all use of the converters of the same type, potentially slowing down things further.
- Input strings are passed by reference as well as most output strings. This is expensive AND eliminates any savings being called with string_view input
- Byte buffer is malloced
- Error handling is limited to rejecting bad codepoints, or rejecting the entire conversion.
  
I did not want to do major surgery to such a critical piece of code at this time, but I did make a version for StringUtils' use. UnicodeConverter is heavily based on CharSetConverter with these major changes:

- Only converts what StringUtils needs: to/from utf8, wchar_t, u32string
- Rather than having one global table of converters, protected by locks, UnicodeConverter uses thread-local tables of converters. This avoids the locking and resource contention issues that CharSetConverter has
- Changed to use string_view as input and string return value (no pass by reference)
- Changed to use char[] buffer from stack instead of malloc
- Three different actions can be taken on conversion error:

1. return empty string  (available in CharSetConverter)
2. Omit bad characters  (available in CharSetConverter)
3. Replace bad characters with "substitution characters" a technique commonly used in browers to show characters that it doesn't recognize.
      
 I have written tests for performance and it shows:

- For strings shorter than 25 chars, UnicodeConverter takes about 47% of the time to convert than CharSetConverter.
- For strings about 500 code-units the gains drop to about 37%.
- At about 800 - 2,500 code units the performance is on par with CharsetConverter.
 
 **FoldCase API**
 Rather than pass the string to FoldCase as IN_OUT by-reference argument, as is done for ToLower, I chose the current recommendation to pass input by string_view and output by string return value. This is both more efficient and easier to read. It also allows for chaining of calls. A fair amount of the performance improvement in UnicodeConverter is due to this API model
 
 About 450 instances of calls to ToLower were changed to call FoldCase

- Each call required changing from pass by reference to pass by string_view and return by value.
- Tedious, but generally straight-forward
- Frequently eliminated temporary string variables
-  Able to chain a number of calls, but many more could be reworked to do that
 
 **ALTERNATIVE CHOICES**
 The choice to modify 450 calls to FoldCase is a lot to review, and not without risk. If this proves too much, I am (reluctantly) willing to go an alternate route. The main thing is that you MUST make all 450 changes (plus other changes) before the Turkish I problem is solved.

1. I could have given FoldCase an API like ToLower so that only the name of the function would need changing. This would have considerably simplified the effort, but would be worse for the long-term, being slower, harder to read and against current style
2. OR I could have created FoldCase with two apis: one following the model as currently implemented, and the other following the ToLower model. This would allow the quick     conversion over to FoldCase from 1, above, but also allow for the conversion to the preferred API over subsequent PRs.

If it is not acceptable to keep with the way I implemented FoldCase, I much prefer option 2, above.
   
**Modification to *NoCase Functions**
All *NoCase functions had to use FoldCase in order to protect from Turkish I problem, etc. There are not so many of these calls and most required changes to the function and not the callers.
  
**ToUpper/ToLower**
At a minimum, an optional locale argument needed to be added to ToUpper/ToLower. Specifying the locale is needed in the few cases where FoldCase can not be used, but protection against Turkish I problem required. Almost always this involved ToUpper. By using C locale non-ASCII characters are not modified by ToUpper. Sometimes English locale can work as.

There were other significant problems:

-  C++ limitations greatly reduced accuracy of converting multi-byte UTF8 via ToUpper or ToLower. This is due to C++ tolower operates a byte at a time, not codepoint. 
- While UTF-8 used C++ built-in conversion tables, Stingutils had it's own private tables for wstring. These tables are very similar to the FoldCase mapping tables, only 10-years (or so) older. Using separate tables meant that string and wstring case conversions would not be identical. And be a bit out of sync with the FoldCase tables
- Chose to use CaseMap.txt conversion table from Unicode.org. Implemented Character Map similar to FoldCase. 
- Supports 500 or so more characters than the old tables used for wstring
- Now have ToLower & ToUpper functions based on these UTF-32 tables producing consistent results for UTF8, wstring and u32string and across platforms
- Chose to change ToLower & ToUpper to use FoldCase API and eliminated old APIS. My justification for this radical move is:

1. The old API harmed performance, readability and usage
2. string_view is flexible, able to accept string_view, string or char*
3. Adding another signature might solve part of the problem, but the addition would clutter and confuse things further
4. Removing the old API made it much easier to be sure that I had modified all instances of the functions
5. Keeping FoldCase, ToLower and ToUpper consistent is valuable
6. About 450 uses of ToLower were changed to FoldCase, leaving about 80 uses of ToLower/ToUpper to modify to use Locale, or to remain as before, but with new API
      
**ALTERNATIVE CHOICEs for ToLower/ToUpper API Change**
I would have made the same API change in order to ensure that I modified all appropriate instances of ToLower/ToUpper. Only by breaking all of the old calls could I be sure that I got them all. Having done that there is still at least one alternative:

- Could add-back signature(s) for old API to ToLower/ToUpper (without any uses of it). While this might make it easier for those familiar with the current API to write code, it may also encourage incorrect code. Forcing an API change forces developers to take a second look at the documentation    
**TESTING**

- Added numerous tests for new and old apis
- **Temporarily** moved old StringUtils to utils/test along with the old TestStringUtils.   Instrumented old TestStringUtils to call new version to help confirm compatability. Cloned new TestStringUtils, modifying to call old StringUtils to help confirm compatability.
- Performance tests for UnicodeConverter, 
- Tests to validate that FoldCase and CaseMapping tablse built correctly
- Using a RandomTrailers addon that I am working on, created fake movies for 5,400 Japanese, Turkish, Chinese, Thai, Russian, etc. movies by downloading movie info from TMDb and creating a dummy movie with the appropriate name for Kodi to scrape
- Manually exercised Kodi using different languages
- Obviously test coverage is small
- Can read some German, but that is about it

**Future plans (subsequent PRs)**
Modify callers of CompareNoCase to switch to using faster or more accurate functions. CompareNoCase requires entire string to be converted to u32string as well as case folded. 

- Most uses of CompareNoCase should be EqualsNoCase, which a length check can frequently resolve, or a partial FoldCase
- Uses of CompareNoCase with length argument are much better to use StartsWithNocase
- Modify StringUtils APIs to use string_view and return by function value instead of by reference. (In order to take advantage of string_view, the call stack needs to use string_view.)
- Modify CharSetConverter with changes made to UnicodeConverter, eliminating UnicodeConverter when complete
- Resume effort to use ICU4C library by getting it to build on all platforms, continuing my work with Android. ICU4C library is big, but can be made much smaller by omitting what you don't need via configuration.

fbacher committed:
Doxy documentation and comments only

NO CODE CHANGES

fbacher commented:
Got behind merging with master.
Development complete.
Awaiting feedback

neo1973 commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -327,96 +203,422 @@ std::wstring StringUtils::FormatV(const wchar_t *fmt, va_list args)
+      }
+    }
+  }
+  return C_LOCALE;
```
Isn't this prone to data races? I think the best option to synchronize this is to initialize the `std::locale` in a lambda like this:
```c++
static const std::locale C_LOCALE = []() {
  // everything from the "if (!initialized)" block
}();
```
Not the `()` at the end to directly invoke the lambda. This also solves the problem that in case of a failure the initialization is repeated over and over again.

yol commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -1894,7 +2188,4455 @@ const std::locale& StringUtils::GetOriginalLocale() noexcept
+    CaseMap(U'\x10cad', U'\x00000', U'\x10ced', U'\x00000'),
+    CaseMap(U'\x10cae', U'\x00000', U'\x10cee', U'\x00000'),
+    CaseMap(U'\x10caf', U'\x00000', U'\x10cef', U'\x00000'),
+    CaseMap(U'\x10cb0', U'\
```
this will make a copy of the whole table... please use a pointer or similar

fbacher replied:
Recoded all of the tables to use constexpr.

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+    return *this;
+  }
+
+  std::string GetSourceCharset(void) const { return m_sourceCharset; }
```
please remove the `void` from `(void)` everywhere

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+  NumberOfStdConversionTypes /* Dummy sentinel entry */
+}; // StdConversionType
+
+// This is where I went nuts creating templates. I am brand-new to Templates
```
please keep in mind that your code is going into a large code base shared by many contributors and will probably live for many years, so it might make sense to save the comments about your personal skills for the PR description or commit message. In the code, they are a bit misplaced :)

fbacher replied:
I wrote them for reviewers. Now is a good time to strip them. Next time I'll figure out how to annotate code using PR tools.

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+
+    static const char* BytePointer(STR_T str)
+    {
+      return static_cast<const char*>(static_cast<const void*>(str.data()));
```
be honest and just use reinterpret_cast in these cases ;-)

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+  /* Ensures the char buffers fed to iconv are cast to the correct
+   * type for the platform.
+   *
+   * iconv may declare inbuf to be char** rather than const char** depending
```
why not just always cast to char**?

fbacher replied:
This came from CharSetConverter and looked nasty, so I left it alone.
But after playing with it  a while I fail to see how we can avoid  doing essentially the same thing. If you always cast to char** then calls to any iconv implementation that  requires const char** will break, such is the one in tools/target/depends:
`size_t iconv (iconv_t cd,
const char* * inbuf, size_t * inbytesleft,
char* * outbuf, size_t * outbytesleft);`

Or am I missing something here? The iconv used on Linux does not declare inbut with a const:
`extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
		     size_t *__restrict __inbytesleft,
		     char **__restrict __outbuf,
		     size_t *__restrict __outbytesleft);`

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+
+      if (errno == E2BIG) // Should only occur for Windows, since buffer size
+      { // is constexpr
+        if (TRACE_OUTPUT) // TODO: remove before merge. Helps with debugging on
```
fix TODO

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+        {
+          iconvRC = 0; // Don't fail on invalid char
+
+          if (substituteBadCodepoints and
```
```suggestion
          if (substituteBadCodepoints &&
```
do not use alternative operators

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+         typename IN_CHAR_T, // = typename ST_IN_T_ARG::value_type,
+         typename OUT_CHAR_T> //= typename ST_OUT_T_ARG::value_type>
+
+ST_OUT_T CUnicodeConverter::CInnerUCodeConverter::convert(
```
this function needs a thorough review; it is security critical and implementing this is prone to edge case problems, off-by-ones etc.
unfortunately, I don't have enough time to do it at the moment. maybe someone else can pitch in?

fbacher replied:
I assume that I should not mark this resolved since you are asking others to continue reviewing.

yol reviewed (changes requested):
Thanks for the extensive work. Unfortunately, I can't do an in-depth review of the functionality changes, but I had a 1 h look through the source code. I hope you can address the points and we can merge this, since I think improving our locale capabilities (even when we don't go the "full way" yet and use ICU) is very useful to a lot of people.

Also, fingers crossed that someone else pitches in for review as well :)

yol commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+  ~CUnicodeConverterType();
+
+  iconv_t PrepareConverter();
+  CUnicodeConverterType& operator=(const CUnicodeConverterType& other)
```
ah good point, so either disallow copying, or implement it properly. also applies to the copy constructor. You could use https://github.com/xbmc/xbmc/blob/master/xbmc/utils/ScopeGuard.h I guess.

fbacher replied:
Good catch. I need to pay more attention. I changed it to:
CUnicodeConverterType& operator=(const CUnicodeConverterType& other) = delete;
Not used in current code

fbacher commented:
I'm working on the suggestions. May take a week or so

fbacher commented on xbmc/utils/StringUtils.h:
```diff
@@ -88,50 +102,529 @@ class StringUtils
+   *        UTF8 format
+   * \return a string containing the result
+   */
+  static std::string ToUtf8(const std::u32string_view str);
```
I thought marking it const mght help the compiler just as marking any local variable const would.
Change made (not committed yet)

fbacher replied:
I made change in latest push. And to think that I liked working on compilers (ancient history)

fbacher commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -9,11 +9,14 @@
-//  Modified to use J O'Leary's std::string class by kraqh3d
+//  Authors:   Frank Feuerbacher, Paul J. Weiss and numerous others
 //
+//  Modified:  To support J O'Leary's std::string class by kraqh3d
```
Shall I revert it?

yol replied:
yeah I guess so ^^;

yol commented:
Just as heads-up: When the review changes are done, we need to have this tested on more platforms before merging

fbacher commented:
When I fix problems, do I make a new commit, or do I rebase the change into the commit in which the original code came from? 

Can this mess up any ongoing reviews (line numbers, lost info)?

yol commented:
I'm fine either way, but before merging it will have to be rebased so we have a clean history. GitHub keeps the old review comments attached to the old code and you can continue to add new comments there, so nothing is really lost. During the review, it's usually easier to see what you changed if you keep new changes as separate commits for now, but I won't insist, so you can do it the way you find it easier.

fbacher commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+      }
+      else
+      {
+        CLog::Log(LOGERROR, "{}: Unknown character type\n", __FUNCTION__);
```
This started as a sanity check, but it is not needed. To test, I replaced with:

else if constexpr (std::is_same<STR_OUT_T, STR_OUT_T_ARG>::value)
          static_assert("STR_OUT_T_ART is Invalid type");
The constexpr can only be true if the enable_if_basic_string does its job
The else if can only be reached if the previous if's did not cover all of the possible returned values (provided all possible values were used as arguments). Any way, the static assert did not reveal invalid type and I know that all three possible supported string types go through this.

Therefore I am removing the sanity test.

yol replied:
Hm I don't really agree, having sanity checks is great. Let's rather have too many than too few. Especially when they are compile-time and thus don't cost any performance. It's great that it everything checks out now, but having these kinds of checks makes it more robust in the long run. So I really think you should keep it :)

fbacher replied:
Change in latest push

fbacher commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -327,96 +203,422 @@ std::wstring StringUtils::FormatV(const wchar_t *fmt, va_list args)
-  wchar_t* p = (wchar_t*) bsearch (&c, unicode_lowers, sizeof(unicode_lowers) / sizeof(wchar_t), sizeof(wchar_t), compareWchar);
-  if (p)
-    return *(unicode_uppers + (p - unicode_lowers));
+  std::wstring result;
```
I think that this should be done selectively.

For all but Windows the stack-based output buffer is sized to contain the same number of code-units (characters) as input + 512 pad bytes. This should normally be more than enough for the conversion, so the conversion loop should normally be one iteration. The output string will be appended to once.

For DARWIN, there is a wrinkle. DARWIN is set to use UTF-8-MAC, which causes iconv to convert and normalize the string. The particular normalization used only expands the string. The amount expands depends upon the characters. Languages like Korean tend to expand more than Latin script based languages. 

For lack of any better guidance, I'll double the stack buffer size for Darwin when encoding
is UTF-8.

Windows c++ compiler forces us to use a fixed-size output buffer which is set to be enough for a 2K u32string. Fortunately I can pre-allocate the output string to be the correct size.

yol replied:
Yeah please apply it where it makes sense (like here). It might be a better idea to use alloca or another solution (e.g., use fixed stack buffer and fall back to heap if the string is too big) instead of dynamic-sized arrays on the stack since this is not a C++ standard feature.

fbacher replied:
I like using memory from stack, but there are warnings about blowing the stack. Should at least add a size limit. It is non-standard but g++ and clang both take it. Using alloc or _alloc (Windows) begins to sound more complicated.

It looks like I could use the result string directly as a buffer- std::string.data(). 

- use reserve to preallocate for expected size, which should be exact (except for Darwin & UTF-8-MAC which can produce longer strings that input, but not shorter). 
- use reserve to grow, if needed
- perhaps use resize_to_fit when done

Another option is thread-local malloc buffers, of limited size, but must worry about leaks.
 
I think either stack (with size limit) or using the result string as the buffer is the way to go

fbacher committed:
StringUtils Mods

fbacher commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -433,61 +635,97 @@ void StringUtils::ToCapitalize(std::wstring &str)
+  // The Simple Unicode support that is supplied here allows a quick
+  // binary comparison of the two strings without conversion.
+
+  if (str1.length() != str2.length())
```
I think it best to have StringUtils function. I think consistently using StringUtils for unicode strings is best. Mixing and matching may work, but could be confusing. Further, the unicode support could change, especially if ICUC4 library is used.

yol replied:
I don't see how your comment relates to the line of code that I highlighted? Maybe misplaced comment?

fbacher replied:
I was being dense. Removed check.

fbacher commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -1894,7 +2188,4455 @@ const std::locale& StringUtils::GetOriginalLocale() noexcept
+    CaseMap(U'\x001fe', U'\x00000', U'\x001ff', U'\x00000'),
+    CaseMap(U'\x001ff', U'\x001fe', U'\x00000', U'\x001fe')};
+
+static const std::vector<CaseMap> CASEMAP_0x0001(
```
Now using C arrays

fbacher commented on xbmc/utils/ExecString.cpp:
```diff
@@ -166,7 +166,7 @@ bool CExecString::Parse(const std::string& execString)
 
   // Keep original function case in execstring, lowercase it in function
-  StringUtils::ToLower(m_function);
+  m_function = StringUtils::FoldCase(m_function);
```
I am not against adding some of the old signatures back. Removing them greatly simplified make sure that I reviewed and made necessary modifications for every occurrence.

I did find that there are by far more cases where a function return is the most useful. There are occasions where passing by IN-OUT is useful, but not a ton of them. I'm always worried about too many overloads confusing things. That said it is time to finalize this, but I'll do it as a separate commit. I'll get back with a proposal of what to change in the next few days.

fbacher commented on xbmc/utils/UnicodeConverter.h:
```diff
@@ -0,0 +1,138 @@
+private:
+  static const int m_Utf8CharMinSize;
+  static const int m_Utf8CharMaxSize;
+  class CInnerUCodeConverter;
```
I copied from CharacterSetConverter, but I can clean up. This is marked private, but I'll see if I can make it more private (not listed at all).

fbacher commented on xbmc/utils/UnicodeConverter.cpp:
```diff
@@ -0,0 +1,992 @@
+// Both logically equivalent, but for byte comparisons you want to use the same
+// representation everywhere.
+
+#define WCHAR_IS_UCS_4 1
```
Next drop will use constexpr every place that I can.

fbacher commented:
Several TODO items

ICUC4 tables that provided the case mapping data have a license agreement (complete agreement in StringUtils.cpp):

 * Copyright © 1991-2022 Unicode, Inc. All rights reserved.
 * Distributed under the Terms of Use in https://www.unicode.org/copyright.html.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of the Unicode data files and any associated documentation
 * (the "Data Files") or Unicode software and any associated documentation
 * (the "Software") to deal in the Data Files or Software
 * without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, and/or sell copies of
 * the Data Files or Software, and to permit persons to whom the Data Files
 * or Software are furnished to do so, provided that either
 * (a) this copyright and permission notice appear with all copies
 * of the Data Files or Software, or
 * (b) this copyright and permission notice appear in associated
 * Documentation.

Two other TODOs that I think should be done as separate PRs once this is complete:

The VAST majority of uses of CompareNoCase can be changed to use EqualsNoCase, which is 
faster since no conversion to Unicode is required. Further, if string length is fixed,
 then a simple length check is all that is needed.
 
 All of the cases that use 'n' should change to StartsWithNoCase or EndsWithNoCase
 because they will be faster and less error-prone.

yol commented:
Hey @fbacher, thanks for all your hard work on this. Unfortunately, I will be very busy this month with reworking our infrastructure (you might have heard about our problems...), so I have little time left to spend on this PR. I'm sorry in advance if that means you have to do more rebasing :( Maybe @lrusak can you take a look here in the meantime?

yol commented:
You also need to fix the builds: https://jenkins.kodi.tv/job/BuildMulti-PR/26262/

fbacher commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -327,96 +203,422 @@ std::wstring StringUtils::FormatV(const wchar_t *fmt, va_list args)
+  const std::locale cLocale = StringUtils::GetCLocale();
 
-  return c;
+  for (auto c : str)
```
Should be resolved in latest push

fbacher commented:
I am trying to resolve the Windows and OSX test case issues. I added some debug tracing in UnicodeConverter, but since kodi-test blocks or redirects stdout from kodi during full test runs I don't get the debug trace in a Jenkins build/test run. Is there some configuration that I can change to fix this? I only need to run several tests from TestCUnicodeConverter.

I do need to get a modern Windows environment of my own. Perhaps this is the time. I'll either make my Linux box dual boot or dive into VMware. As for as OSX, I might convince my wife to let me use one of her several Macs for some testing. She is very protective of them.

fbacher commented:
Somehow force push did not push all changes made to some files. Made minor changes and re-pushed.

fbacher commented:
Resolved one error in UnicodeCharconverter where it overran input limit.
Made some test improvements and now going through getting them to work across Windows and DARWIN, which have their quirks.

fbacher commented:
It seems that sometimes jenkins/git has trouble getting the right code. It looks fine here, but when I look at a Console's log it shows errors with line numbers off by a dozen or more. I'll force another push and see what happens.

fbacher commented:
I should have converted to draft some time back to reduce the noise to everyone. 
I should have fixed within week. Mostly a bit more verification as well as cleanup of some debugging code

ulville commented:
Hi @fbacher ! First of all, thanks for your great efforts on fixing this gangrened issue for good 🙏 I don't mind using Kodi in English but I also set it up for my family on an old laptop connected to the kitchen tv and my mom sometimes gets lost in the menus. And I know a lot of people just doesn't know what's causing this issue and just assumed Kodi is broken for good in the last updates. In my opinion this issue really hurts the reputation of Kodi for Turkish speaking people which are apprx. 100 million around the world. And I can't even imagine the frustration of the translators which their work is just useless now. So please excuse my hurrying.

What is the current state of this? Are we close to the fixed version?

fbacher commented:
> Hi @fbacher ! First of all, thanks for your great efforts on fixing this gangrened issue for good pray I don't mind using Kodi in English but I also set it up for my family on an old laptop connected to the kitchen tv and my mom sometimes gets lost in the menus. And I know a lot of people just doesn't know what's causing this issue and just assumed Kodi is broken for good in the last updates. In my opinion this issue really hurts the reputation of Kodi for Turkish speaking people which are apprx. 100 million around the world. And I can't even imagine the frustration of the translators which their work is just useless now. So please excuse my hurrying.
> 
> What is the current state of this? Are we close to the fixed version?

I'm trying to get this into the next release. I've had a few interruptions along the way. It is a BIG change so will take time for everyone to review, make changes and sign off.

This is NOT the ultimate fix, but should make things significantly better. After this is complete I want to go back to adding icu4c to Kodi, but that requires getting the library integrated into every platform build. It will be the following release before that can be done. Fortunately, the changes here can be largely reused with icu4c lib underneath. After the switch to icu4c, then other changes can be made, incrementally.

fbacher commented:
As far as I know I have resolved all issues. 
Sorry for the long absence, I had several interruptions. I'll be on vacation for about a week.

fbacher commented:
The currently open change request is actually a request for others to review a section. I am not marking it as resolved until I get further feedback
